rust/otel-arrow-rust/src/encode/record/array.rs

```diff
@@ -259,32 +477,111 @@ where
 
     /// Append a value to the underlying builder
     fn append_value(&mut self, value: &T) {
-        self.initialize_inner();
-        let inner = self
-            .inner
-            .as_mut()
-            .expect("inner should now be initialized");
-        match inner {
-            MaybeDictionaryBuilder::Native(native_builder) => {
-                native_builder.append_value(value);
-            }
-            MaybeDictionaryBuilder::Dictionary(dictionary_builder) => {
-                match dictionary_builder.append_value(value) {
-                    Ok(_) => {}
-                    Err(DictionaryBuilderError::DictOverflow {}) => {
-                        let mut native = TN::new(self.inner_args.clone());
-                        dictionary_builder.to_native(&mut native);
-                        self.inner = Some(MaybeDictionaryBuilder::Native(native));
-                        self.append_value(value);
-                    }
-                }
-            }
-        }
+        // temporarily move the value of default_value to avoid borrowing self as mut and non-mut
+        // at the same time
+        let default_value = std::mem::take(&mut self.default_value);
+        let is_default = || Self::is_default_value(&default_value, value);
+        self.handle_append(
+            is_default,
+            &default_value,
+            |native| native.append_value(value),
+            |dict| dict.append_value(value),
+            |me| me.append_value(value),
+        );
+
+        // restore the default value
+        self.default_value = default_value;
```

Comment:
Interesting. Do you think:
(a) the compiler is going to eliminate the take() and restore
(b) it would be worse to make a copy and drop it
(c) all of the above
(d) none of the above?